src/app/data-model-provider/MetadataList.cpp

```diff
@@ -25,106 +25,158 @@ namespace app {
+    if (mBuffer == nullptr)
+    {
+        mBuffer            = static_cast<uint8_t *>(Platform::MemoryCalloc(numElements, mElementSize));
+        mCapacity          = numElements;
+        mBufferIsAllocated = true;
+        return mBuffer != nullptr ? CHIP_NO_ERROR : CHIP_ERROR_NO_MEMORY;
+    }
 
-    VerifyOrReturnError(mBuffer != nullptr, CHIP_ERROR_NO_MEMORY);
+    // we already have the data in buffer. we have two choices:
+    //   - allocated buffer needs to be extended
+    //   - re-used const buffer neext to be copied over
+    if (mBufferIsAllocated)
+    {
+        auto new_buffer = static_cast<uint8_t *>(Platform::MemoryRealloc(mBuffer, (mElementCount + numElements) * mElementSize));
+        VerifyOrReturnError(new_buffer != nullptr, CHIP_ERROR_NO_MEMORY);
+        mBuffer = new_buffer;
+    }
+    else
+    {
+        // this is NOT an allocated buffer, but it should become one
+        auto new_buffer    = static_cast<uint8_t *>(Platform::MemoryCalloc(mElementCount + numElements, mElementSize));
+        mBufferIsAllocated = true;
+        VerifyOrReturnError(new_buffer != nullptr, CHIP_ERROR_NO_MEMORY);
+        memcpy(new_buffer, mBuffer, mElementCount * mElementSize);
+        mBuffer = new_buffer;
+    }
+    mCapacity = mElementCount + numElements;
 
-    mAllocated = true;
-    mCapacity  = numElements;
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR GenericMetadataList::AppendRaw(const void * buffer)
+CHIP_ERROR GenericAppendOnlyBuffer::AppendSingleElementRaw(const void * buffer)
 {
-    VerifyOrReturnError(!mIsImmutable, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mElementCount < mCapacity, CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(mElementCount < mCapacity, CHIP_ERROR_BUFFER_TOO_SMALL);
     memcpy(mBuffer + mElementCount * mElementSize, buffer, mElementSize);
     mElementCount++;
     return CHIP_NO_ERROR;
 }
 
-void GenericMetadataList::Invalidate()
+CHIP_ERROR GenericAppendOnlyBuffer::AppendElementArrayRaw(const void * buffer, size_t numElements)
 {
-    if ((mBuffer != nullptr) && mAllocated)
+    ReturnErrorOnFailure(EnsureAppendCapacity(numElements));
+
+    memcpy(mBuffer + mElementCount * mElementSize, buffer, numElements * mElementSize);
+    mElementCount += numElements;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericAppendOnlyBuffer::ReferenceExistingElementArrayRaw(const void * buffer, size_t numElements)
+{
+    if (mBuffer == nullptr)
     {
-        chip::Platform::MemoryFree(mBuffer);
-        mBuffer    = nullptr;
-        mAllocated = false;
+        // we can NEVER append with 0 capacity, so const cast is safe
+        mBuffer       = const_cast<uint8_t *>(static_cast<const uint8_t *>(buffer));
+        mElementCount = numElements;
+        // The assertions below are because we know the buffer is null/not allocated yet
+        // Not included because these seem to affect flash size surprisingly enough...
```

Comment:
What does "not included" mean?

Reply:
Deleted ... I had `assert()` before and that increases flash. When I made VerifyOrDie the increase became negligible (on qpg at least), so I left them in ... seems better to check than silently fail. Good catch!